### ISSUE TITLE
Add an async OnceCell and Lazy

### ISSUE THREAD

**Issue by Johannesd3** [tokio-rs/tokio#3482]
If I want to make e.g. an http request once, and use the result during the rest of the program, the `once_cell` crate would be a good solution in synchronous Rust. But I found no way to hand over an async fn. So I'd like to have an async counterpart of `OnceCell` and `Lazy`.

The async `OnceCell<T>` should have a function like `fn get_or_init(f: impl FnOnce() -> impl Future<Output = T>) -> impl Future<Output = &T>`. If the OnceCell is already initialized, the future resolves immediately. If there's an ongoing initialization, it waits until it's ready. Otherwise it runs the given `FnOnce` and waits for the future to finish.

The async `Lazy` takes such a closure on creation and executes it the first time the value is requested, similar to `Lazy` in `once_cell`.

I tried to implement it myself using existing (safe) components, but got no satisfactory solution. 

Maybe there's another easy solution to solve this problem. Or there's a pitfall why an async `OnceCell` could not work. At least I did not find previous discussions about this topic. I also hope this in the scope of this project, but since there are many other such things in `tokio::sync`, I thought it's a good place to ask.

**Comment by Darksonn** [comment]
I have thought about this a few times, and I think it would be a nice idea to have. It is kinda tricky to do properly, but it can be done.

**Comment by b-naber** [comment]
I can try to implement this. I have one question though. In `get_or_init` we presumably have to use a mutex in case `OnceCell` is not already initialized. If `f` itself tries to call `get_or_init`, we would deadlock. Does anybody know whether there's a way to avoid this or is it sufficient to warn users in the docs not to call `get_or_init` again inside `f`?

**Comment by Darksonn** [comment]
The non-async `OnceCell` panics if `get_or_init` is called inside of itself, but I don't know if there's a good way to detect that. It is fine to deadlock, although it should of course be mentioned in the documentation.

**Comment by b-naber** [comment]
I have some questions about the API for this. [Here's my current implementation](https://github.com/b-naber/tokio/blob/change_once_cell_api/tokio/src/sync/once_cell.rs)

Currently I'm not able to a return a reference to the value in the `OnceCell` from `get_or_init`. Compared to the sync `OnceCell` implementation in `once_cell`, where a static instance of `OnceCell` is created and shared refs can be spawned on threads, I settled for allowing `OnceCell` to be wrapped in `Arc`s in order to spawn them. The reason for doing this was that I needed a `Pin<&mut Self>` in order to implement that `get_or_init` waits when an initialization is already going on. So I decided to store an `Arc<Inner>` in `OnceCell` and create a Future, which has a handle to `Inner`, to implement the 'wait while initializing' feature. But then we can't e.g. do the following:

```

#[tokio::main]
async fn main() {
    tracing_subscriber::fmt::init();
    let span = span!(Level::DEBUG, "main");
    let _enter = span.enter();

    //let once: OnceCell<u32> = OnceCell::new();
    let once = Arc::new(OnceCell::new());
    let once_clone1 = once.clone();
    let once_clone2 = once.clone();
    let handle1 = tokio::spawn(async move {
        once_clone1.get_or_init(call_async_fn).await
    });
    let handle2 = tokio::spawn(async move {
        once_clone2.get_or_init(call_async_fn).await
    });

    let result1 = handle1.await;
    let result2 = handle2.await;
}
```

Instead we currently would need to call `get` again once we successfully awaited the spawned `get_or_init` call in order to get access to a reference of the value in `OnceCell`.

```

#[tokio::main]
async fn main() {
    tracing_subscriber::fmt::init();
    let span = span!(Level::DEBUG, "main");
    let _enter = span.enter();

    //let once: OnceCell<u32> = OnceCell::new();
    let once = Arc::new(OnceCell::new());
    let once_clone1 = once.clone();
    let once_clone2 = once.clone();
    let handle1 = tokio::spawn(async move {
        once_clone1.get_or_init(call_async_fn).await;
    });
    let handle2 = tokio::spawn(async move {
        once_clone2.get_or_init(call_async_fn).await;
    });

    handle1.await;
    handle2.await;
    let result = once.get();
    if let Some(ref v) = result {
        println!("result: {:?}", v);
    }
}
```

So that `get_or_init` should really be called something like `try_init`, which we can await and once it returns we can call `get()` on the OnceCell.

Is this a dealbreaker? 

I'd really appreciate if someone more experienced could take a look at the implementation. This is the first async code I've written and I have likely missed something that's obvious to someone more experienced.

**Comment by Johannesd3** [comment]
I think `Arc` is a dealbreaker: `new` should be a `const fn` in order to allow `static OnceCell`s, but with `Arc` this is not possible. Did you try to give `OnceCellResult` a lifetime parameter instead, so it can't live longer than its `OnceCell`? But I'm not sure, I don't know much about these things.

**Comment by Darksonn** [comment]
I haven't looked at the impl in detail, but that implementation has just one `Waker`. You can't do that since there may be multiple others waiting for the value.

**Comment by Darksonn** [comment]
I think it could be implemented like this:
```Rust
struct OnceCell<T> {
    value_set: AtomicBool,
    value: UnsafeCell<MaybeUninit<T>>,
    lock: tokio::sync::Semaphore,
}
```
The value would be initialized to the following, which can be done in a const way.
```Rust
OnceCell {
    value_set: AtomicBool::new(false),
    value: UnsafeCell::new(MaybeUninit::uninit()),
    lock: Semaphore::const_new(1),
}
```
Then to access the value, you perform the following steps:

 1. Check `value_set` with acquire ordering. If `true`, we are done and can read `value`.
 2. Acquire a permit from the semaphore.
 3. If acquire fails because the semaphore is closed, then `value_set` is now guaranteed to be `true`. We are done and can read `value`.
 4. If a permit is acquired, start running the initialization function.
 5. If the initialization function is cancelled (e.g. `select!`) or panics, the destructor of `SemaphorePermit` releases the permit, and some other task can acquire it instead.
 6. If the initialization function succeeds, then _first_ set the value, _then_ update `value_set` with release ordering and _then_ close the semaphore. And in this order.

**Comment by b-naber** [comment]
@Darksonn Great, it works. Very elegant solution, thanks a lot.

> I haven't looked at the impl in detail, but that implementation has just one Waker. You can't do that since there may be multiple others waiting for the value.

I thought about that initially, but when reading the [docs on Future](https://doc.rust-lang.org/std/future/trait.Future.html#return-value) it said:

"Note that on multiple calls to poll, only the Waker from the Context passed to the most recent call should be scheduled to receive a wakeup."

I mistakenly thought that this statement holds for any `poll` call on any `Future` instance of `OnceCellResult` (but that doesn't really make any sense^^). I don't understand how `Context` works, it's totally mysterious to me. Can you explain what happens behind the scenes with `Context` on a `poll` call?

**Comment by Darksonn** [comment]
@b-naber Feel free to submit a PR that implements the description.

Regarding the multiple wakers thing, the problem is if _multiple_ calls are made to `get_or_init`, each call will produce a separate `Future`. You need to store a separate `Waker` for each future, but only one per future, even if there are multiple calls to `poll` on the futures.

**Comment by Darksonn** [comment]
An `OnceCell` was added in #3591. I don't think we will add a `Lazy` type to Tokio because automatically computing the value in a `Deref` impl is not possible as you can't use `.await` inside the `Deref` impl.

**Comment by the8472** [comment]
> An `OnceCell` was added in #3591. I don't think we will add a `Lazy` type to Tokio because automatically computing the value in a `Deref` impl is not possible as you can't use `.await` inside the `Deref` impl.

Imo some kind of memoizing wrapper would still be useful even if it takes an extra method call or `await` on every use.
E.g. for expensive precomputations that are tossed into a separate Task and the results will later be used multiple times.  Afaict there isn't anything like that at the moment.

**Comment by Darksonn** [comment]
I recommend using the pattern shown in this example in the docs:
```rust
use tokio::sync::OnceCell;

static ONCE: OnceCell<u32> = OnceCell::const_new();

async fn get_global_integer() -> &'static u32 {
    ONCE.get_or_init(|| async {
        1 + 1
    }).await
}

#[tokio::main]
async fn main() {
    let result = get_global_integer().await;
    assert_eq!(*result, 2);
}
```

**Comment by the8472** [comment]
The advantage of Lazy (or LazyLock in std) is that it takes a closure which makes it more convenient to capture state that's used for the initialization and then stuffing the lazy value into to a struct.

**Comment by NobodyXu** [comment]
@Darksonn Having a `Lazy` would definitely be useful to cargo-binstall, currently we have to use `OnceCell` to emulate `Lazy` https://github.com/cargo-bins/cargo-binstall/blob/zeroize/crates/binstalk/src/helpers/lazy_gh_api_client.rs

```rust
use std::{future::Future, sync::Mutex};

use binstalk_git_repo_api::gh_api_client::GhApiClient;
use tokio::sync::OnceCell;
use zeroize::Zeroizing;

use crate::{
    errors::BinstallError,
    helpers::{remote, tasks::AutoAbortJoinHandle},
};

pub type GitHubToken = Option<Zeroizing<Box<str>>>;

#[derive(Debug)]
pub struct LazyGhApiClient {
    client: remote::Client,
    inner: OnceCell<GhApiClient>,
    task: Mutex<Option<AutoAbortJoinHandle<GitHubToken>>>,
}

impl LazyGhApiClient {
    pub fn new(client: remote::Client, auth_token: GitHubToken) -> Self {
        Self {
            inner: OnceCell::new_with(Some(GhApiClient::new(client.clone(), auth_token))),
            client,
            task: Mutex::new(None),
        }
    }

    pub fn with_get_gh_token_future<Fut>(client: remote::Client, get_auth_token_future: Fut) -> Self
    where
        Fut: Future<Output = GitHubToken> + Send + Sync + 'static,
    {
        Self {
            inner: OnceCell::new(),
            task: Mutex::new(Some(AutoAbortJoinHandle::spawn(get_auth_token_future))),
            client,
        }
    }

    pub async fn get(&self) -> Result<&GhApiClient, BinstallError> {
        self.inner
            .get_or_try_init(|| async {
                let task = self.task.lock().unwrap().take();
                Ok(if let Some(task) = task {
                    GhApiClient::new(self.client.clone(), task.await?)
                } else {
                    GhApiClient::new(self.client.clone(), None)
                })
            })
            .await
    }
}
```

**Comment by NobodyXu** [comment]
There's [another place](https://github.com/cargo-bins/cargo-binstall/blob/main/crates/detect-targets/src/desired_targets.rs) where we have to use `Arc<OnceCell>` as well to simulate `Lazy`:

```rust
use crate::detect_targets;

use std::sync::Arc;

use tokio::sync::OnceCell;

#[derive(Debug)]
enum DesiredTargetsInner {
    AutoDetect(Arc<OnceCell<Vec<String>>>),
    Initialized(Vec<String>),
}

#[derive(Debug)]
pub struct DesiredTargets(DesiredTargetsInner);

impl DesiredTargets {
    fn initialized(targets: Vec<String>) -> Self {
        Self(DesiredTargetsInner::Initialized(targets))
    }

    fn auto_detect() -> Self {
        let arc = Arc::new(OnceCell::new());

        let once_cell = arc.clone();
        tokio::spawn(async move {
            once_cell.get_or_init(detect_targets).await;
        });

        Self(DesiredTargetsInner::AutoDetect(arc))
    }

    pub async fn get(&self) -> &[String] {
        use DesiredTargetsInner::*;

        match &self.0 {
            Initialized(targets) => targets,

            // This will mostly just wait for the spawned task,
            // on rare occausion though, it will poll the future
            // returned by `detect_targets`.
            AutoDetect(once_cell) => once_cell.get_or_init(detect_targets).await,
        }
    }

    /// If `DesiredTargets` is provided with a list of desired targets instead
    /// of detecting the targets, then this function would return `Some`.
    pub fn get_initialized(&self) -> Option<&[String]> {
        use DesiredTargetsInner::*;

        match &self.0 {
            Initialized(targets) => Some(targets),
            AutoDetect(..) => None,
        }
    }
}

/// If opts_targets is `Some`, then it will be used.
/// Otherwise, call `detect_targets` using `tokio::spawn` to detect targets.
///
/// Since `detect_targets` internally spawns a process and wait for it,
/// it's pretty costy, it is recommended to run this fn ASAP and
/// reuse the result.
pub fn get_desired_targets(opts_targets: Option<Vec<String>>) -> DesiredTargets {
    if let Some(targets) = opts_targets {
        DesiredTargets::initialized(targets)
    } else {
        DesiredTargets::auto_detect()
    }
}
```

**Comment by Darksonn** [comment]
I would usually expect that simulating `Lazy` would be done like this:
```rs
use tokio::sync::OnceCell;

static ONCE: OnceCell<u32> = OnceCell::const_new();

async fn get_global_integer() -> &'static u32 {
    ONCE.get_or_init(|| async {
        1 + 1
    }).await
}

#[tokio::main]
async fn main() {
    let result = get_global_integer().await;
    assert_eq!(*result, 2);
}
```
but your examples seem significantly more complicated. Is there a reason for the other design?

**Comment by NobodyXu** [comment]
> Is there a reason for the other design?

I want the future to be executed in parallel, using `tokio::spawn`.

The code above spawns one or more processes, and on some environments they might be quite expensive to execute.

**Comment by Darksonn** [comment]
The `get_or_init` method on Tokio's `OnceCell` will not start the closure more than once. The first call will start execution, and any future calls will wait for the first call to execute.

**Comment by NobodyXu** [comment]
Yes I know that, but I want the future to be executed in parallel, which is why I added extra complexity to it.

**Comment by ettolrach** [comment]
`LazyCell` and `LazyLock` will get stabilised in Rust 1.80. Will we get a tokio equivalent, or should we stick with `tokio::sync::OnceCell`?

**Comment by Darksonn** [comment]
The current plan is situation we will not have `Lazy` and that you should stick with `OnceCell`. We can't mirror the API in std because we can't run async code inside `Deref`. Even if add a `get` method, I recall there being some limitations with async closures in global variables.

If there is a proposal for how we could add a `Lazy` to Tokio, then we could reconsider the decision. But the proposal would need to show that it's actually useful for real-world use-cases (that is, the limitations have been fixed or can be worked around).
